### PR TITLE
add unit tests for `pl.utilities.grads`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds tests and coverage for `pytorch_lightning.utilities.grads.grad_norm` function.
Found an edge case where L0 "norm" (p=0) does not yield the correct result for the total norm. This is because the implementation of the total norm there only works for proper norms (the L0 is not a real norm). 

Option A: special case the L0 norm computation
Option B: don't support `norm_type=0`

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/utilities/test_grads.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import torch
+import torch.nn as nn
+
+from pytorch_lightning.utilities import grad_norm
+
+
+@pytest.mark.parametrize(
+    "norm_type,expected",
+    [
+        # bug for L0 "norm"! total norm should be 5 here
```
Comment (carmocca) [inline tests/utilities/test_grads.py:26]: Why do you say it is a bug?

```python
np.linalg.norm(np.array([3., 2.]), 0)
Out[31]: 2.0
```

which matches the definition of vector 0-norm: `sum(x != 0)`

Comment (awaelchli) [reply tests/utilities/test_grads.py:26]: yes, 100% a bug in our implementation, I am sure. Agree your example is fine and I'm not saying `torch.norm` computes an incorrect result. It's our implementation that takes a shortcut for efficiency to avoid recomputing values when computing the total  norm of all parameters combined. That shortcut holds up for p>0 because of properties of the Lp-norm. But since for p=0 it is not a true norm, that formula breaks down. Here is what we do:

Let v0 and v1 be two vectors. 

Individual norms:
```
n0 = norm(v0, p) 
n1 = norm(v1, p)
```

Total norm of all elements combined:
```
t = norm([v0, v1], p)
```

Instead of the above, our implementation does this using the already computed values n0 and n1:
```
t = norm([n0, n1], p)
```

One can show that
```
norm([v0, v1], p) = norm([n0, n1], p) for p > 0
```

But for p = 0 this does not hold true. 
Example:

```
v0 = [-1, 2]
v1 = [3, 4]

```
We will get `n0 = n1 = 2`
but
```
2 = norm([2, 2], 0) != norm([-1, 2, 3, 4], 0) = 4
```

Comment (carmocca) [reply tests/utilities/test_grads.py:26]: This hasn't been an issue because:

https://github.com/PyTorchLightning/pytorch-lightning/blob/44aed17affc396a82ea3bb93345c1892ebaee82b/pytorch_lightning/loops/optimization/optimizer_loop.py#L485

but if we don't want to support 0-norm, it should raise an exception instead of silently skipping it
